Implementing embeddings

defeo opened the pull request:
This branch is still incomplete. Pushing for comments.

defeo commented:
So, I finally managed to have a complete implementation of the naive algorithms. This should make a good starting point to play with in Nemo, while I port the non-naive algorithms (Allombert's). This branch is still missing

- [ ] More thorough tests,
- [ ] Documentation.

In the meanwhile, I'd be grateful if someone could make a code review: I had to hack around quite a bit with Filint's "templates" to avoid repeating myself. The most dubious hacks include:

- Use of macros and `static __inline__` functions in `.c` template files to mock missing functions in `fmpz_*` or `nmod_*` (see `matrices.c`);
- Implementation of a very basic, incomplete, header-only, `fmpz_mod_mat_t` type. This should maybe be a separate PR.

Note that moving to C++ wouldn't help much here: it would certainly be more succinct, but the real problem is that the base types do not have as uniform an interface as one may hope.

I'm eager to know what you think of the code.

defeo commented:
I just documented all functionality, except the new wrapper `fmpz_mod_mat`. Not sure what to do with that.

wbhart commented:
I will eventually look at and comment on this. At present, I am really time poor and lots of people are annoyed that I haven't done various things. I hope that by the end of the year I can catch back up with everything. It's been one of the busiest years ever for me.

defeo commented:
No worries, Bill, we're in no rush. I take the occasion to welcome @erou in the project (he's advancing on the Nemo side), and to thank him for stress-testing this branch.

Note: I'm hunting down a segfault in `fq_zech_embed`. So far it's been quite mysterious, I'm not even convinced this code is at fault. I won't have much time for it these next two weeks, though.

defeo commented:
Note: tests do not pass because this PR is waiting for a resolution to #419.

wbhart commented:
Of course we need to get the tests passing again. I think the current failures are due to the MPFR library having again changed, rather than anything you are doing. I'll try to fix that now, if I can.

wbhart commented:
Ah, no, actually, the test failures are segfaults in your new code. This will need to be fixed. I recommend valgrinding it first on a 64 bit machine, then examining the code to see if you can spot the problem. Otherwise we'll need to find a suitable machine to run the code on to debug it in the normal way.

wbhart commented:
Ah, ok. I could try to get Dan to work on that. Or do you have plans to
sort that out yourself?

On 18 January 2018 at 15:58, Luca De Feo <notifications@github.com> wrote:

> Note: tests do not pass because this PR is waiting for a resolution to
> #419 <https://github.com/wbhart/flint2/issues/419>.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/wbhart/flint2/pull/351#issuecomment-358671637>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAOzpig5Z9b5NJQUWf5xVGWrCzw9-FBGks5tL1wlgaJpZM4Or0sy>
> .
>

defeo commented:
> Ah, ok. I could try to get Dan to work on that. Or do you have plans to
> sort that out yourself?

The thing is that I do not know what to do with it. Maybe we should
continue the discussion in #419.

wbhart commented:
Please ignore the approval. The user who is responsible has been reported for abuse.

defeo commented:
!! How was this possible?!

Anyway, now that I'm done with other stuff, it seems a right time to come back to this PR. I shall update it soon.

wbhart commented:
I honestly don't know how it was possible. He started to claim he had control over my repos, but I changed my password, proving that he did not.

tkoolen commented:
Anybody can review a PR, you don't need special permission for that. The account has also opened 141 PRs like the ones you've been seeing (also https://github.com/JuliaArrays/StaticArrays.jl/pull/499).

wbhart commented:
@defeo @fieker has been asking about this implementation. I have completely lost track of everything here. From your paper, this is naieve embeddings with BCS right, but with our Conway polynomial finite fields?

Is it worth us inviting you again late this year to have another push to finish this and the associated Nemo PR off?

What is still outstanding from our side that would need to be finished before we could merge this (say before any potential visit from you)?

We also noticed there is a C implementation in your repo, which I presume is an Allombert implementation? Is this likely to be useful? Is the plan still to put something like that in Flint?

Sorry for being away from this topic for so long. I am trying to cover too many bases, as usual.

wbhart commented:
@defeo Actually, I guess we never got as far as a Nemo PR. At least I don't see it any more if it existed. Do you have a branch somewhere? I didn't see any branch in your version of the Nemo repo which sounds like embeddings.

defeo commented:
Hi Bill!

> From your paper, this is naive embeddings with BCS right, but with our Conway polynomial finite fields?

Just BCS, works with any polynomials, including Conway.

@erou completed this work and presented it at ISSAC last year. His package is here: https://github.com/erou/Nemo-embeddings-demo

The normal course of action would be to update this PR with the C code in Édouard's repo, and open another PR to Nemo.

> Is it worth us inviting you again late this year to have another push to finish this and the associated Nemo PR off?

I think it is, but I'm basically not available before November. However Édouard may be interested in visiting, if you can agree on a date.

> We also noticed there is a C implementation in your repo, which I presume is an Allombert implementation? Is this likely to be useful? Is the plan still to put something like that in Flint?

We've been doing a lot of research on this, which is the reason why I haven't made a PR containing Allombert's algorithm. We have a paper appearing at ISSAC: https://github.com/erou/gf-h90-lattice (soon to be on arXiv), with code: https://github.com/erou/LatticeGFH90.jl

This is much better than BCS in many ways: memory requirements linear in the number of stored fields and their degrees, faster embedding computation, decently fast embedding evaluation (although for small fields storing matrices is faster). However, there is one big drawback: it requires Conway polynomials... so you may as well just use Conway polynomials!

We're still thinking about ways of making this construction useful in practice. I'm positive this is possible, however I'm afraid it would be rather involved. It could be an interesting project to look at how this could be integrated into Nemo.

> Sorry for being away from this topic for so long. I am trying to cover too many bases, as usual.

No worries. It's the same for me, sorry about that.

wbhart commented:
Édouard is welcome to come. We have an Oscar developer's meeting June 11-14. He's of course welcome to come to that, although it may be better to schedule a later week in which we can devote more full attention to him. I am not available Aug 18-23, but that's about it so far.

erou commented:
Hi Bill,

I am not available until September, but I would be happy to come in Autumn!

Here is [the article](https://arxiv.org/abs/1906.00870) Luca was talking about. The code of this article should work, but our last (very hacky) work https://github.com/erou/Nemo-embeddings-demo presented in ISSAC last year is no longer working on Binder.

wbhart commented:
@erou That sounds great. I've made a note to invite you (we will pay) a bit closer to the time. We aren't sure of our commitments just yet, but we'll certainly be in contact in plenty of time.

wbhart commented:
@erou We said that we were going to invite you in Autumn for a visit for a week. Are you by any chance available from Oct 14-18? I just discussed this with @fieker and certainly next month would be slightly better than this month, but we are still very keen to have you visit.

erou commented:
@wbhart Oct 14-18 is totally fine!

wbhart commented:
@erou Fantastic. I'll be in contact by email as soon as I can.

fieker commented:
On Wed, Sep 18, 2019 at 09:28:03AM -0700, wbhart wrote:
> @erou Fantastic. I'll be in contact by email as soon as I can.

Great!
> 
> -- 
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub:
> https://github.com/wbhart/flint2/pull/351#issuecomment-532762293

wbhart commented:
Thanks @defeo

defeo commented:
Hello,

In prepartaion for Édouard's visit, we cleaned up the code, fixed the merge conflict, and addressed some of Bill's comments. There are a couple of items on which I'm still doubtful, see below:

>     * fq_mul_matrix probably should be fq_embed_mul_matrix, since function names should follow the name of the module.

Ok. But why do you single out `fq_mul_matrix`? There are `fq_trace_matrix`, `fq_compostion_matrix_sub`, `fq_composition_matrix`, `fq_mono_to_dual_matrix`, `fq_dual_to_mono_matrix` too. Shouldn't those be changed too?

>     * I personally don't like double underscore methods like __fmpz_mod_inv_degree. Putting double underscores on functions that are super secret is not a good solution. The functions should either be given a name which is descriptive enough that no one will use them accidentally, or they should be put in the correct location elsewhere in Flint, for general use (if appropriate)
 
But those double underscore functions are either macros or static inlines in C files. How could they be called accidentally? I understand that the macros `__[fmpz_|n]mod_poly_get_coeff` could be useful elsewhere and thus we could add them to the respective header files and document them. But `__[fmpz|_n]mod_inv_degree` is an internal hack with no other use whatsoever.

wbhart commented:
Regarding the names, yes, if they are functions in a given module, they should be named after the module. That's the convention. I probably just overlooked the others.

Regarding the underscores, I should have explained. Underscores in Flint are reserved for other purposes, e.g. identifying lower level layers of a module, e.g. fmpz_poly_mul vs _fmpz_poly_mul.

The convention in Flint for function hiding is just to use static functions if they are in one file only, or put helper in the name or otherwise just pick a name that is meaningful that won't accidentally be called.

wbhart commented on fq_embed/doc/fq_embed.txt:
```diff
@@ -0,0 +1,158 @@
+
+    Convert \code{a} to a column vector of length \code{degree(ctx)}.
+
+void fq_set_fmpz_mod_mat(fq_t a, const fmpz_mod_mat_t col, const fq_ctx_t ctx)
```
Can we move this to the fq module. It looks generic enough.

wbhart commented on fq_embed/doc/fq_embed.txt:
```diff
@@ -0,0 +1,158 @@
+    Convert a column vector \code{col} of length \code{degree(ctx)} to
+    an element of \code{ctx}.
+
+void fq_composition_matrix(fmpz_mod_mat_t matrix,
```
possibly change to fq_embed_composition_matrix? And same for functions below.

wbhart commented on fq_embed/get_set_fmpz_mod_mat.c:
```diff
@@ -0,0 +1,35 @@
+
+#include "fq_embed.h"
+
+void fq_get_fmpz_mod_mat(fmpz_mod_mat_t col,
```
Move to fq?

wbhart commented on fq_embed/get_set_fmpz_mod_mat.c:
```diff
@@ -0,0 +1,35 @@
+        fmpz_zero(fmpz_mod_mat_entry(col, i, 0));
+}
+
+void fq_set_fmpz_mod_mat(fq_t a,
```
And this?

wbhart commented on fq_embed/mul_matrix.c:
```diff
@@ -0,0 +1,53 @@
+            if (i > 0)
+                fmpz_sub(fmpz_mod_mat_entry(matrix, i, j),
+                         fmpz_mod_mat_entry(matrix, i, j),
+                         fmpz_mod_mat_entry(matrix, i-1, j-1));
```
spaces before and after -

wbhart commented on fq_embed/mul_matrix.c:
```diff
@@ -0,0 +1,53 @@
+        for (i = 0; i < len; i++)
+        {
+            fmpz_mul(fmpz_mod_mat_entry(matrix, i, j),
+                     fmpz_mod_mat_entry(matrix, len-1, j),
```
spaces

wbhart commented on nmod_mat.h:
```diff
@@ -153,6 +157,12 @@ FLINT_DLL void nmod_mat_neg(nmod_mat_t B, const nmod_mat_t A);
 FLINT_DLL void nmod_mat_scalar_mul_add(nmod_mat_t dest, const nmod_mat_t X,
                                 const mp_limb_t b, const nmod_mat_t Y);
+NMOD_MAT_INLINE
+void nmod_mat_scalar_mul_fmpz(nmod_mat_t res, const nmod_mat_t M, const fmpz_t c)
```
This looks a bit dodgy. What if fmpz is large? Shouldn't there be a reduction?

wbhart commented on fq_zech_embed/doc/fq_zech_embed.txt:
```diff
@@ -0,0 +1,158 @@
+      matrix.
+    \end{itemize}
+
+void fq_zech_trace_matrix(nmod_mat_t res, const nmod_mat_t basis,
```
Similar comments for function names in this module.

wbhart commented on fq_nmod_embed/mul_matrix.c:
```diff
@@ -0,0 +1,47 @@
+    for (i = gen->length; i < len; i++)
+        nmod_mat_entry(matrix, i, 0) = 0;
+
+    /* M[i,j] = M[i-1,j-1] - M[len-1,j-1] * lead * ctx->modulus->coeffs[i] */
```
spacing

wbhart commented on fq_nmod_embed/mul_matrix.c:
```diff
@@ -0,0 +1,47 @@
+    /* M[i,j] = M[i-1,j-1] - M[len-1,j-1] * lead * ctx->modulus->coeffs[i] */
+    for (j = 1; j < len; j++) {
+        nmod_mat_entry(matrix, len-1, j) = 
+            nmod_mul(nmod_mat_entry(matrix, len-1, j-1), lead, mod);
```
spacing

wbhart commented on fq_nmod_embed/mul_matrix.c:
```diff
@@ -0,0 +1,47 @@
+            nmod_mul(nmod_mat_entry(matrix, len-1, j-1), lead, mod);
+        for (i = 0; i < len; i++) {
+            nmod_mat_entry(matrix, i, j) = 
+                nmod_mul(nmod_mat_entry(matrix, len-1, j), 
```
spaces

wbhart commented on fq_nmod_embed/mul_matrix.c:
```diff
@@ -0,0 +1,47 @@
+            if (i > 0)
+                nmod_mat_entry(matrix, i, j) =
+                    nmod_sub(nmod_mat_entry(matrix, i, j),
+                             nmod_mat_entry(matrix, i-1, j-1),
```
spaces

wbhart commented on fq_embed_templates/test/t-embed_matrices.c:
```diff
@@ -0,0 +1,141 @@
+
+    /* Check random emebedding (degrees 1..5) */
+    for (j = 1; j < 6; j++) {
+        for (i = 0; i < (6-j) * flint_test_multiplier(); i++)
```
spacing

wbhart commented on fq_embed/mul_matrix.c:
```diff
@@ -0,0 +1,53 @@
+    for (i = gen->length; i < len; i++)
+        fmpz_zero(fmpz_mod_mat_entry(matrix, i, 0));
+
+    /* M[i,j] = M[i-1,j-1] - M[len-1,j-1] * lead * ctx->modulus->coeffs[i] */
```
spacing

wbhart commented on fmpz_mod_mat.h:
```diff
@@ -0,0 +1,297 @@
+    fmpz_mat_scalar_mul_fmpz(B->mat, A->mat, c);
+    _fmpz_mod_mat_reduce(B);
+}
+
```
We should add fmpz_mod_mat_scalar_mul_fmpz_mod

wbhart commented on fmpz_mod_mat.h:
```diff
@@ -0,0 +1,297 @@
+/*
```
We should add documentation and test code and add it as an actual module.

wbhart commented on nmod_mat.h:
```diff
@@ -153,6 +157,12 @@ FLINT_DLL void nmod_mat_neg(nmod_mat_t B, const nmod_mat_t A);
 FLINT_DLL void nmod_mat_scalar_mul(nmod_mat_t B, const nmod_mat_t A, mp_limb_t c);
```
These need documentation.

wbhart commented:
Yay!
